### PR TITLE
allow to set specific settings editors + add test

### PR DESCRIPTION
**Pull request by 3nids**
This a better approach to support custom settings editor widgets (this is the remaining part of https://github.com/qgis/QGIS/pull/59006, and follow-up of https://github.com/qgis/QGIS/pull/59014)

You can now specifiy a custom settings editor for a specific setting.

This also adds a test for the settings editor registry both for the automatic and custom cases.

### CODE REVIEW

File: src/gui/settings/qgssettingstreemodel.cpp [hunk 1]
```diff
@@ -486,7 +486,10 @@ void QgsSettingsTreeItemDelegate::setEditorData( QWidget *editor, const QModelIn
 {
   QgsSettingsEditorWidgetWrapper *eww = QgsSettingsEditorWidgetWrapper::fromWidget( editor );
   if ( eww )
-    eww->setWidgetFromVariant( index.model()->data( index, Qt::DisplayRole ) );
+  {
+    QVariant value = index.model()->data( index, Qt::DisplayRole );
```
Comment (nirvn) [inline src/gui/settings/qgssettingstreemodel.cpp:490]:
```suggestion
    const QVariant value = index.model()->data( index, Qt::DisplayRole );
```

File: src/gui/settings/qgssettingseditorwidgetregistry.h [hunk 2]
```diff
@@ -45,14 +45,21 @@ class GUI_EXPORT QgsSettingsEditorWidgetRegistry
...
+     */
+    void addWrapperForSetting( QgsSettingsEditorWidgetWrapper *wrapper SIP_TRANSFER, const QgsSettingsEntryBase *setting SIP_KEEPREFERENCE );
+
     //! Returns a new instance of the editor widget for the given \a id
-    QgsSettingsEditorWidgetWrapper *createWrapper( const QString &id, QObject *parent ) const;
+    QgsSettingsEditorWidgetWrapper *createWrapper( const QString &id, QObject *parent ) const SIP_FACTORY;
 
     //! Creates an editor widget for the given \a setting using the corresponding registered wrapper
-    QWidget *createEditor( const QgsSettingsEntryBase *setting, const QStringList &dynamicKeyPartList, QWidget *parent = nullptr ) const SIP_FACTORY;
+    QWidget *createEditor( const QgsSettingsEntryBase *setting, const QStringList &dynamicKeyPartList, QWidget *parent = nullptr ) const SIP_KEEPREFERENCE;
```
Comment (nyalldawson) [inline src/gui/settings/qgssettingseditorwidgetregistry.h:58]: Is keep reference correct here? I suspect this should be SIP_TRANSFERBACK, for the same reasons as described here: https://github.com/qgis/QGIS/blob/master/src/core/processing/qgsprocessingalgorithm.h#L83

Comment (3nids) [reply src/gui/settings/qgssettingseditorwidgetregistry.h]: I tried, the test passes and I could successfully edit settings from the plugin. So all good on the practical side!

**Review by nirvn — APPROVED** [review]

**Comment by 3nids** [comment]
Postgres test failure seems unrelated and is currently happening on other PRs.

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**